## Morales, Santiago
## Torres , Juan Manuel

In [16]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog

##Punto 1


In [17]:
c = np.array([-3,-5])

A = np.array([[1,1],
              [2,4],
              [1,2],
              [-1,0],
              [0,-1]])

b = np.array([500,1000,2000,-100,-150])

bounds = [(  100, None),
              (150, None),
              ]

res = linprog(c, A_ub=A, b_ub=b,
                  bounds=bounds, method='revised simplex')

res

<ipython-input-17-061703751a91>:15: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res = linprog(c, A_ub=A, b_ub=b,


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -1350.0
       x: [ 2.000e+02  1.500e+02]
     nit: 2

## La producción óptima del producto A es de 200 y la del producto B es de 150, de esta forma, la ganancia total sería de $1.350

## Punto 2

In [18]:
c2 = np.array([0.1 , 0.15 , 0.16 , 0.08])

A2 = np.array([[0.5 , 0.3 , 0.25 , 0.6],
              [0.3 , 0.1 , 0.4 , 0.2],
              [0.2 , 0.6 , 0.35 , 0.2],
              ])

b2 = np.array([35 , 30 , 35])

C = np.array([
               [1 , 1 , 1 , 1],
              ])

D = np.array([100])




bounds2 = [(  0, None),
            (0, None),
          (0, None),
          (0, None),
              ]

res2 = linprog(-c2, A_ub=A2, b_ub=b2, A_eq=C, b_eq=D,
                  bounds=bounds2, method='revised simplex')

res2

<ipython-input-18-4cc2c7d93d37>:25: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res2 = linprog(-c2, A_ub=A2, b_ub=b2, A_eq=C, b_eq=D,


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -13.73684210526316
       x: [ 0.000e+00  1.579e+01  5.789e+01  2.632e+01]
     nit: 5

In [19]:
import cvxpy as cp
n=4
x = cp.Variable(n)
obj = cp.Minimize(-c2@x)
restricciones = [A2@x <= b2,  C@x >= D, x[0:3]>=0]
prob = cp.Problem(obj, restricciones)
prob.solve()

# Print result.
print("El valor óptimo es", prob.value)
print(round(pd.DataFrame(x.value),2))


El valor óptimo es -13.73684210144203
       0
0   0.00
1  15.79
2  57.89
3  26.32


## La distribución de los 100 millones en los 4 portafolios es: 0 millones para el primer portafolio, 15.79 millones para el segundo portafolio, 57.89 millones para el tercer portafolio, y 26.32 millones para el cuarto portafolio, de esta forma, el retorno esperado sería de 13.74  

## Punto 3

In [21]:
X = np.array([0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 1])

A3 = np.array([[1 , 0 , 0 , 0 , 0 , 1 , 0 , 0 , -1 , 0 , 0 , 0 , 0 , 0],
               [-1.01 , 1 , 0 , 0 , 0 , 0 , 1 , 0 , 1.003 , -1 , 0 , 0 , 0 , 0],
               [0 , -1.01 , 1 , 0 , 0 , 0 , 0 , 1 , 0 , 1.003 , -1 , 0 , 0 , 0],
               [0 , 0 , -1.01 , 1 , 0 , -1.02 , 0 , 0 , 0 , 0 , 1.003 , -1 , 0 , 0 ],
               [0 , 0 , 0 , -1.01 , 1 , 0 , -1.02 , 0 , 0 , 0 , 0 , 1.003 , -1 , 0],
               [0 , 0 , 0 , 0 , -1.01 , 0 , 0 , -1.02 , 0 , 0 , 0 , 0 , 1.003 , -1]])

b3 = np.array([150 , 100 , -200 , 200 , -50 , -300])

bounds3 = [(0 , 100),
           (0 , 100),
           (0 , 100),
           (0 , 100),
           (0 , 100),
           (0 , None),
           (0 , None),
           (0 , None),
           (0 , None),
           (0 , None),
           (0 , None),
           (0 , None),
           (0 , None),
           (0 , None),]

res3 = linprog(-X, A_eq=A3, b_eq=b3,
                  bounds=bounds3, method= 'revised simplex')
res3 .x

<ipython-input-21-cc7b279817fc>:27: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res3 = linprog(-X, A_eq=A3, b_eq=b3,


array([  0.        ,  50.98039216,   0.        ,   0.        ,
         0.        , 150.        ,  49.01960784, 203.43436358,
         0.        ,   0.        , 351.9441675 ,   0.        ,
         0.        ,  92.49694915])

In [22]:
import cvxpy as cp
n=14
x = cp.Variable(n)
obj = cp.Minimize(-X@x)
restricciones = [A3@x >= b3, x[0:5] <=100, x[0:13] >= 0]
prob = cp.Problem(obj, restricciones)
prob.solve()



-92.49694914810405

## Al maximizar la función, el saldo del sexto mes (Junio) es de $92.4969

##Punto 4

In [34]:

c4 =np.array([0, 0, 0, 0, 0, 0, 0, 1])

R = (1+0.0001)**7

A4 = np.array([[1, 0, 0, 0, -1, 0, 0, 0],
              [0, 1, 0, 0, R, -1, 0, 0],
              [0, 0, 1, 0, 0, R, -1, 0],
              [0, 0, 0, 1, 0, 0, R, -1]])

b4 = np.array([[150, -100, -50, 200]])


res4 = linprog(-c4, A_ub=A4, b_ub=-b4, method='simplex')
pd.DataFrame(res4.x).round(2).T


<ipython-input-34-b1d1a2cd62d8>:13: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res4 = linprog(-c4, A_ub=A4, b_ub=-b4, method='simplex')


0    1    2    3      4      5     6       7
0  0.0  0.0  0.0  0.0  150.0  50.11  0.14  200.14

##Punto 5


In [23]:

c5 = np.array([0, 0, 0, 5, 8, 10])


A5 = np.array([[0, 0, 0, 1, 0, 0],
                [0, 1, 0, 0, 0, 0],
                [0, 0, 1, 0, 0, 0],
                [0, 0, 0, 0, 1, 0],
                [0, 0, 0, 0, 0, 1]])

b5 = np.array([2000, 300, 250, 2000, 2000])

C2 = np.array([[1, 1, 1, 0, 0, 0],
                  [0, 0, 0, 1, 1, 1]])

D2 = np.array([1000, 10000])


res5 = linprog(-c5, A_ub = -A5, b_ub = -b5, A_eq = C2, b_eq = D2, method = 'simplex')
res5


<ipython-input-23-9c2c4d2096e3>:18: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res5 = linprog(-c5, A_ub = -A5, b_ub = -b5, A_eq = C2, b_eq = D2, method = 'simplex')


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -86000.0
       x: [ 4.500e+02  3.000e+02  2.500e+02  2.000e+03  2.000e+03
            6.000e+03]
     nit: 3

## La función objetivo maximizada es igual a $86.000 ; y el presupuesto de ventas es de 2000, el de marketing es de 2000, y el de producción es de 6000

##Punto 6